PSF image representation in NDData/CCDData

parejkoj opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

This pull request adds a simple PSF image to NDData/CCDData, to support the Rubin Observatory LSST Prompt Processing alert distribution system. The LSST alert packet image cutouts will be sent as [astropy CCDData objects](https://jira.lsstcorp.org/browse/RFC-696), and the only thing we see as missing from that format is the PSF. For our purposes for these image cutouts, all we need is a realization of the PSF image at the image center, not a generic parameterized or model PSF.

TODO:

* Are any further tests or functionality necessary?
* Does this serialize to the various astropy outputs, or do I need to add code for that?
* Document how to make use of this PSF representation (but I'm not sure where such documentation would go).
* Document how to use the astropy flux-preserving warping code to move the PSF around within the image?

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

pllim commented on astropy/nddata/nddata.py:
```diff
@@ -118,7 +121,7 @@ class NDData(NDDataBase):
     meta = MetaData(doc=_meta_doc, copy=False)
 
-    def __init__(self, data, uncertainty=None, mask=None, wcs=None,
+    def __init__(self, data, uncertainty=None, mask=None, wcs=None, psf=None,
```
Is it safer to put new keyword in the end in case people been passing in keywords as pos args?

parejkoj replied:
I was actually wondering if we shouldn't modify this API to make everything after `data` be a keyword-only, since there's a lot of them? So it would look like: `__init__(self, data, *, uncertainty=None, ...)`.

mwcraig replied:
> While adding this, I wondered if we shouldn't modify this API to make everything after data be keyword-only, since there's a lot of them? 

I don't think we want to do this in minor version bump; there may be issues also with the `@use_nddata` decorator but I don't recall the details of that.

pllim commented:
Thanks! Definitely need a change log.

Maybe a What's New too but I'll let others decide on that one.

mwcraig reviewed (changes requested):
@parejkoj -- this looks good. Can you please add an example in the narrative documentation?

It also needs a changelog, and I think it warrants a "What's new" entry too.

One question: why not just shove the PSF into the metadata?

parejkoj commented:
> One question: why not just shove the PSF into the metadata?

That is one of the things I was wondering about: I think it hinges on the question of how much structure you want to have in the CCDData object. Having a dedicated name for the PSF makes it readily discoverable, and we can enforce the type of the object. LSST's equivalent to CCDData (`afw.Exposure`) is a C++ object underneath, so it needs to have a pretty rigid structure and contain explicit types and names for all that "metadata".

I can think of a handful of other things one might add as named objects on CCDData (e.g. an LSF for spectra), but none of them are as broadly useful as the PSF, to my mind. Eventually extending this PSF to include model representations seems like a future next step, but we don't need it for our purposes, and there's not really any way to generically represent a model PSF. I think the question is "what, besides WCS, should be a first-class named parameter?" I think the PSF is the next thing. An object representing the photometric calibration isn't as broadly applicable as the WCS and PSF are, otherwise it would be my next suggestion.

mwcraig commented:
@parejkoj -- it is best to mention me in replies -- my github notifications are a train wreck because of conda-forge (I know I could fix that, but) (and still my bad for not checking back sooner).

> Having a dedicated name for the PSF makes it readily discoverable, and we can enforce the type of the object. LSST's equivalent to CCDData (afw.Exposure) is a C++ object underneath, so it needs to have a pretty rigid structure and contain explicit types and names for all that "metadata".

That makes sense -- I'm fine with adding it at a fairly high level.

Do you have time to do the changelog and some basics documentation this week?

parejkoj commented:
@mwcraig :
> Do you have time to do the changelog and some basics documentation this week?

I'm trying to finish it up now.

One of the difficulties I've had is figuring out exactly where I should put all the necessary functionality and tests. I'm fixing the one line I missed that the coverage test pointed out, but I don't really know where else I need to add code. Suggestions very welcome there.

parejkoj commented:
@mwcraig : Ok, I believe I've added working FITS write/read support and a test that demonstrates it, there's a note in the changelog and what's new, and I've added some notes to the narrative docs (though I'm not sure what else should go there). Further comments very welcome.

dhomeier commented on astropy/nddata/tests/test_ccddata.py:
```diff
@@ -55,7 +56,7 @@ def test_ccddata_unit_cannot_be_set_to_none():
 def test_ccddata_meta_header_conflict():
     with pytest.raises(ValueError) as exc:
```
```suggestion
    with pytest.raises(ValueError, match=r"can't have both header and meta."):
```
Good catch! But generally we recommend the more convenient `match=` functionality for new tests now
(https://docs.astropy.org/en/latest/development/testguide.html#testing-exceptions); won't need the `as exc` then anymore – the same applies to all the added tests below.

mwcraig replied:
@parejkoj -- let me know if you want me to make the test changes.

parejkoj replied:
@mwcraig yes, if you could make those test changes, that would be useful, since I'm less familiar with pytest.

mwcraig replied:
@dhomeier @parejkoj -- I'm going to open a separate PR in a few minutes to fix all of the nddata tests to use `match`, including the one you suggested, Derek.

I'll revert the change @parejkoj  made at this spot since it is unrelated to the psf functionality.

mwcraig replied:
Edit: I won't bother with the revert, since any conflict can be fixed during the rebase.

dhomeier commented on astropy/nddata/ccddata.py:
```diff
@@ -129,6 +129,9 @@ class CCDData(NDDataArray):
             ``ValueError``
 
+    psf : `numpy.ndarray` or None, optional
+        Normalized PSF image at the center of this image.
```
```suggestion
        Normalized image (2D array) representing the PSF at the center of this image.
```
I'd be a bit more explicit here (if I understood correctly from the PR description); my first thought when reading this was "does this mean the 'PSF image' needs to be centred in some way?"
Without knowing much about how this will be typically used or how it is implemented in the Rubin data, might also be helpful to comment on the above point – i.e. would the PSF still be centred from that input, or could it in fact be offset to produce an image shift? But from the added tests it looks as if the implementation as such does not make particular assumptions on the shape of the PSF, so perhaps that is a bit too specific.

parejkoj replied:
I didn't want to make any assumptions in the code about how a user would implement the PSF image for their specific purposes. I could see some details about that normalization being different for e.g. a radio or x-ray image vs. one from LSST, and I don't think we want to encode any assumptions in the code here. I'll take your docstring suggestion, but remove the `2d array`, since ccddata doesn't assume that `data` is 2d either.

We'll want to have some docs about how to e.g. shift this PSF image around the actual image to use it in measurements, but I think that's out of scope for this PR (we want to be able to start making sample alert packets for LSST, but exactly how they'll be used with astropy code is up in the air).

mwcraig replied:
How about just removing the word "nromalized"? You say [here](https://github.com/astropy/astropy/pull/13743/files#diff-8af8355c2d4da08d8c5f4f2faf410fa23c3d0e4a17f7981c284cd4c3fefa8873R142) that `NDData/CCDdata` don't do anything the PSD except carry it around, so there is not a requirement for normalization.

parejkoj replied:
@mwcraig : I guess we could take the word `normalized` out from here, and add something in the narrative docs (or maybe to a parenthetical here?) to the effect of "the PSF image is typically normalized to sum to unity, to ensure that convolving it with the image is a flux-preserving operation"?

dhomeier reviewed (commented):
Some comments from a relatively outsider point of view.
To elaborate a bit more on the requirements a PSF image has to fulfil, all the docs mention it should be 'normalized', but that does not seem to be checked on input anywhere – at least the somewhat pathological `np.random.normal(size=[20, 20])` used in the tests generally is not (where by 'normalized' I'd understand that convolution with such a PSF will be automatically flux-preserving; if something else is meant, that should probably be pointed out in the docs as well).

mwcraig commented on docs/nddata/ccddata.rst:
```diff
@@ -123,21 +123,24 @@ more details about bit planes and the functions ``astropy`` provides for
 
-WCS
-+++
+WCS and PSF
```
My preference would be to make PSF its own bullet point, partly because the way that WCS and PSF are populated is vary different -- WCS comes from the header and PS from an image extension.

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -297,6 +309,16 @@ def wcs(self, wcs):
+
+    @psf.setter
+    def psf(self, value):
+        if value is not None and not isinstance(value, np.ndarray):
```
Does the requirement that PSF be an image have to made here or it could it be done in `CCDData`? I suppose the PSF could be the table or something, and it isn't until the PSF is written/read from `CCDData`  that it really has to be be an `ndarray`.

My thinking here is to keep the top level `NDData` as general as possible.

parejkoj replied:
I think that until we have a more generic way of representing PSFs, we should ensure that it is an array everywhere. Even the nddata docstring says `image representing the psf`, which I think is what we want? That said, I don't actually know how the nddata vs. ccddata functionality divide is actually intended.

mwcraig reviewed (commented):
This is close to ready to go. One last big question: how should PSF's be handled in arithmetic operations? My inclination is to treat it like metadata, where I think we punt to the user to decide what the resulting metadata should be. The logic for several of the other properties is here: https://github.com/astropy/astropy/blob/main/astropy/nddata/mixins/ndarithmetic.py#L231

parejkoj commented:
@mwcraig : The PSF should definitely not be modified during arithmetic operations. This sort of question is one of the reasons the LSST CCDData-like object (`afw.Exposure`) does not have any arithmetic operations itself, only the "image array" class does (`afw.MaskedImage`, which includes the mask and variance planes and which are properly handled during arithmetic operations).

parejkoj commented:
I believe I've taken care of all of the comments. @mwcraig and I had a chat earlier today to resolve things. I've added a warning to ndarithmetic, more narrative docs, psf to nddatabase, cleaned up the docstrings, and rebased to main.

Should I rebase squash my changes, or leave them as-is?

dhomeier commented on astropy/nddata/ccddata.py:
```diff
@@ -270,18 +285,19 @@ def uncertainty(self, value):
                hdu_flags=None, wcs_relax=True,
-               key_uncertainty_type='UTYPE', as_image_hdu=False):
+               key_uncertainty_type='UTYPE', as_image_hdu=False,
+               hdu_psf="PSFIMAGE"):
```
```suggestion
               hdu_psf='PSFIMAGE'):
```
see below

mwcraig replied:
```suggestion
               hdu_psf='PSFIMAGE'):
```

dhomeier commented on astropy/nddata/ccddata.py:
```diff
@@ -270,18 +285,19 @@ def uncertainty(self, value):
-            Default is ``'MASK'`` for mask, ``'UNCERT'`` for uncertainty and
-            ``None`` for flags.
+            Default is ``'MASK'`` for mask, ``'UNCERT'`` for uncertainty,
+            "PSFIMAGE" for psf, and ``None`` for flags.
```
```suggestion
            ``'PSFIMAGE'`` for psf, and ``None`` for flags.
```
Was this formatted so differently on purpose? The single/double quotes in the function signature above are apparently homogenised automatically in the processes docs, but the way this is rendered in the docstring https://astropy--13743.org.readthedocs.build/en/13743/api/astropy.nddata.CCDData.html#astropy.nddata.CCDData.to_hdu
looks very strange to read.

dhomeier commented on astropy/nddata/ccddata.py:
```diff
@@ -535,7 +556,7 @@ def _generate_wcs_and_update_header(hdr):
 def fits_ccddata_reader(filename, hdu=0, unit=None, hdu_uncertainty='UNCERT',
                         hdu_mask='MASK', hdu_flags=None,
-                        key_uncertainty_type='UTYPE', **kwd):
+                        key_uncertainty_type='UTYPE', hdu_psf="PSFIMAGE", **kwd):
```
```suggestion
                        key_uncertainty_type='UTYPE', hdu_psf='PSFIMAGE', **kwd):
```
Again, see comment above (for these cases I have no idea if the quote question becomes all moot once this is all subjected to `black`, but until then I'd keep the style consistent).

dhomeier commented on astropy/nddata/ccddata.py:
```diff
@@ -669,30 +699,34 @@ def fits_ccddata_reader(filename, hdu=0, unit=None, hdu_uncertainty='UNCERT',
 def fits_ccddata_writer(
         ccd_data, filename, hdu_mask='MASK', hdu_uncertainty='UNCERT',
         hdu_flags=None, key_uncertainty_type='UTYPE', as_image_hdu=False,
+        hdu_psf="PSFIMAGE",
```
1 more

mwcraig replied:
```suggestion
        hdu_psf='PSFIMAGE',
```

dhomeier commented on astropy/nddata/ccddata.py:
```diff
@@ -669,30 +699,34 @@ def fits_ccddata_reader(filename, hdu=0, unit=None, hdu_uncertainty='UNCERT',
-        Default is ``'MASK'`` for mask, ``'UNCERT'`` for uncertainty and
-        ``None`` for flags.
+        Default is ``'MASK'`` for mask, ``'UNCERT'`` for uncertainty,
+        "PSFIMAGE" for psf, and ``None`` for flags.
```
```suggestion
        ``'PSFIMAGE'`` for psf, and ``None`` for flags.
```

dhomeier reviewed (changes requested):
Apart from the comments on formatting, the discussion of normalisation etc. now looks very nice.
One more general point (not specifically for this PR) though is that I found it quite hard to navigate to the detailed documentation in `nddata.rst` and `ccddata.rst` from https://astropy--13743.org.readthedocs.build/en/13743/nddata/index.html, as almost all mentions of the two classes are linking to the API docs instead – only under 'Using ``nddata``' are the narrative doc chapters linked. I think the documentation would benefit a lot if they could get another link e.g. from the respective chapter headings.

mwcraig reviewed (changes requested):
I think we need just one more change (which I'll make in a few minutes) -- I think we need to make `psf` a concrete, not abstract, method in `NDDataBase` to ensure we don't introduce a backwards incompatible change. If we leave it as-is then anyone who currently subclasses NDData will need to implement a concrete `psf`. In our conversation yesterday, @parejkoj and I agreed the intent here was a non-breaking change; it simply hadn't occurred to me until a few minutes ago that leaving it abstract could be an issue.

mwcraig commented:
pre-commit.ci autofix

mwcraig reviewed (approved):
Assuming the CI passes I am ok with merging this. Thanks for getting this going @parejkoj!

@dhomeier -- thanks for your detailed review! I think all of your comments have been addressed, and I've opened an issue about improving the documentation landing page with the intent of fixing it for 6.0.

pllim commented:
@dhomeier requested changes -- does he want to re-review?

pllim commented:
Do the commits need squashing? 😸

dhomeier commented:
Thanks for opening the issue on documentation!

mwcraig commented:
> Do the commits need squashing? 😸

@pllim -- I can do the squash if you want (I don't have strong feelings either way)

pllim commented on docs/changes/nddata/13743.feature.rst:
```diff
@@ -0,0 +1 @@
+Add a PSF image representation to ``astropy.nddata.CCDData`` and ``astropy.nddata.CCDData``.
```
Do you mean this?
```suggestion
Add a PSF image representation to ``astropy.nddata.NDData`` and ``astropy.nddata.CCDData``.
```

pllim commented on docs/nddata/ccddata.rst:
```diff
@@ -138,6 +138,21 @@ The ``wcs`` attribute of a `~astropy.nddata.CCDData` object can be set two ways.
+The ``psf`` attribute should be a normalized image representing the PSF at the center of the `~astropy.nddata.CCDData`, sized appropriately for the data; users are responsible for managing and interpreting it in context.
+For more on normalizing a PSF image, see :ref:`astropy:kernel_normalization`.
+
+The ``psf`` attribute is set to ``None`` in the output of an arithmetic operation, no matter the inputs. A warning message is emitted if either of the input images contain a non-``None`` psf; users are responsible for determining the appropriate thing to do in that context.
```
p.s. Would be nice to break long sentences into shorter lines.

```suggestion
The ``psf`` attribute is set to `None` in the output of an arithmetic operation, no matter the inputs. A warning message is emitted if either of the input images contain a non-`None` PSF; users are responsible for determining the appropriate thing to do in that context.
```

pllim replied:
I have the opposite opinion on this. Splitting one sentence into multiple lines makes partial changes in future PRs easier to review. Though it is annoying if someone ends up moving words up/down the lines. So, there is no foolproof solution.

I don't think there is an official stance but if you look at the rest of the docs, they are all broken up into shorter lines.

mwcraig commented:
Thanks for the review @pllim -- I'll update around noon locally

parejkoj commented on astropy/nddata/ccddata.py:
```diff
@@ -129,6 +129,11 @@ class CCDData(NDDataArray):
             If the unit is ``None`` or not otherwise specified it will raise a
             ``ValueError``
 
+    psf : `numpy.ndarray` or None, optional
```
I believe a psf image should always be unitless, but I'm not positive of that.

pllim replied:
I am just thinking what if someone stored it as data with pixel unit attached to it.

parejkoj replied:
hmmm... I'm not sure what to do on that point: I think the psf image as being purely an array (because the thing you do with it is (de)convolve it with the image), and I can't think of a reason someone would want it to have pixel units or what would be gained by that. Might be worth waiting to see how use-cases shake out and if anyone requests it?

pllim replied:
That is fine. I was just curious if this was discussed.

parejkoj commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -257,6 +257,10 @@ def _arithmetic(self, operation, operand,
 
+        # If both are None, there is nothing to do.
+        if self.psf is not None or operand.psf is not None:
+            warnings.warn(f"Not setting psf attribute during {operation.__name__}.", RuntimeWarning)
```
I didn't even know that was a thing. I don't have a strong opinion either way; your suggestion might make the warning more obvious to astropy users?

pllim replied:
Yes, I do think AstropyUserWarning is more explicit (short of creating your own warning class).

parejkoj commented on astropy/nddata/tests/test_ccddata.py:
```diff
@@ -1081,3 +1082,48 @@ def test_read_write_tilde_paths(home_is_tmpdir):
+def test_ccddata_with_psf():
+    psf = _random_psf.copy()
+    ccd = CCDData(_random_array.copy(), unit=u.adu, psf=psf)
+    assert (ccd.psf == psf).all()
```
I was just following the other tests above.

mwcraig replied:
I *think* the reasoning was that the created `ccd` should have an exact copy of the input `psf` in this case (and the other similar cases) so we really do want to check for exact equality.

parejkoj replied:
Not that it really matters, but `np.testing.assert_array_equal` is exact equality. `np.testing.assert_array_almost_equal` is probably the one you're thinking of.

pllim commented on astropy/nddata/ccddata.py:
```diff
@@ -270,18 +285,19 @@ def uncertainty(self, value):
         Parameters
         ----------
-        hdu_mask, hdu_uncertainty, hdu_flags : str or None, optional
+        hdu_mask, hdu_uncertainty, hdu_flags, hdu_psf: str or None, optional
```
```suggestion
        hdu_mask, hdu_uncertainty, hdu_flags, hdu_psf : str or None, optional
```

pllim replied:
I think this was not applied?

pllim commented on astropy/nddata/ccddata.py:
```diff
@@ -578,6 +599,10 @@ def fits_ccddata_reader(filename, hdu=0, unit=None, hdu_uncertainty='UNCERT',
 
+    hdu_psf : str or None, optional
+        FITS extension from which the psf image should be initialized. If the
+        extension does not exist the psf of the CCDData is ``None``.
```
```suggestion
        extension does not exist the psf of the CCDData is `None`.
```

mwcraig replied:
Sorry about that, I thought I had applied them all and then put the other changes on top of it.

pllim commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -1,7 +1,7 @@
 # This module implements the Arithmetic mixin to the NDData class.
 
-
+import warnings
```
```suggestion
```

pllim commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -257,6 +257,10 @@ def _arithmetic(self, operation, operand,
 
+        # If both are None, there is nothing to do.
+        if self.psf is not None or operand.psf is not None:
+            warnings.warn(f"Not setting psf attribute during {operation.__name__}.", RuntimeWarning)
```
You'll have to add the import yourself. I cannot suggest it at the line where it is supposed to go. Now precommit complains if it is not at the right place.

```suggestion
            warnings.warn(f"Not setting psf attribute during {operation.__name__}.", AstropyUserWarning)
```

pllim commented on astropy/nddata/mixins/tests/test_ndarithmetic.py:
```diff
@@ -1,5 +1,7 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 
+import warnings
+
```
```suggestion
```

pllim replied:
This import not longer needed?

pllim commented on astropy/nddata/mixins/tests/test_ndarithmetic.py:
```diff
@@ -1213,3 +1215,21 @@ def test_arithmetics_unknown_uncertainties():
+    ndd1 = NDDataArithmetic(np.ones((3, 3)), psf=np.zeros(3))
+    ndd2 = NDDataArithmetic(np.ones((3, 3)), psf=None)
+
+    # no warning if both are None
+    with warnings.catch_warnings():
+        warnings.simplefilter("error")
+        ndd2.add(ndd2)
```
I think this still applies.

pllim commented on astropy/nddata/mixins/tests/test_ndarithmetic.py:
```diff
@@ -1213,3 +1215,21 @@ def test_arithmetics_unknown_uncertainties():
+        warnings.simplefilter("error")
+        ndd2.add(ndd2)
+
+    with pytest.warns(RuntimeWarning, match="Not setting psf attribute during add"):
```
```suggestion
    with pytest.warns(AstropyUserWarning, match="Not setting psf attribute during add"):
```

pllim commented on astropy/nddata/mixins/tests/test_ndarithmetic.py:
```diff
@@ -1213,3 +1215,21 @@ def test_arithmetics_unknown_uncertainties():
+
+    with pytest.warns(RuntimeWarning, match="Not setting psf attribute during add"):
+        ndd1.add(ndd2)
+    with pytest.warns(RuntimeWarning, match="Not setting psf attribute during add"):
```
```suggestion
    with pytest.warns(AstropyUserWarning, match="Not setting psf attribute during add"):
```

pllim commented on astropy/nddata/mixins/tests/test_ndarithmetic.py:
```diff
@@ -1213,3 +1215,21 @@ def test_arithmetics_unknown_uncertainties():
+        ndd1.add(ndd2)
+    with pytest.warns(RuntimeWarning, match="Not setting psf attribute during add"):
+        ndd2.add(ndd1)
+    with pytest.warns(RuntimeWarning, match="Not setting psf attribute during add"):
```
```suggestion
    with pytest.warns(AstropyUserWarning, match="Not setting psf attribute during add"):
```

mwcraig commented:
@pllim -- thanks for the review. I think all of your comments have been addressed.

pllim commented on astropy/nddata/tests/test_ccddata.py:
```diff
@@ -1081,3 +1082,48 @@ def test_read_write_tilde_paths(home_is_tmpdir):
+
+    # cannot set with non-ndarray
+    with pytest.raises(TypeError, match="The psf must be a numpy array."):
+        ccd.psf = "something"
```
I think this would be a more realistic test case.
```suggestion
        ccd.psf = _random_array * u.pix
```

mwcraig commented:
Hopefully this time I *really* have them all!

pllim reviewed (approved):
LGTM now. Thanks!

pllim commented:
Interesting... Could it be that you actually support Quantity by accident?

```
_______________________________ test_psf_setter ________________________________

    def test_psf_setter():
        psf = _random_psf.copy()
        ccd = CCDData(_random_array.copy(), unit=u.adu)
        ccd.psf = psf
        assert (ccd.psf == psf).all()
    
        # cannot set with non-ndarray
>       with pytest.raises(TypeError, match="The psf must be a numpy array."):
E       Failed: DID NOT RAISE <class 'TypeError'>

.../astropy/nddata/tests/test_ccddata.py:1104: Failed
```

mwcraig commented:
> Interesting... Could it be that you actually support Quantity by accident?
🤔 maybe? Will check this evening but will be away from the keyboard for a few hours now

mwcraig commented:
Turns out quantity is supported (though we don't get credit for that):

```python
In [1]: from astropy import units as u

In [3]: import numpy as np

In [4]: foo = 5 * u.pixel

In [5]: foo
Out[5]: <Quantity 5. pix>

In [6]: isinstance(foo, np.ndarray)
Out[6]: True

mwcraig commented on astropy/nddata/tests/test_ccddata.py:
```diff
@@ -1081,3 +1082,48 @@ def test_read_write_tilde_paths(home_is_tmpdir):
+
+    # cannot set with non-ndarray
+    with pytest.raises(TypeError, match="The psf must be a numpy array."):
+        ccd.psf = _random_array * u.pix
```
```suggestion
        ccd.psf = 5
```

pllim replied:
But... do you really *want* to support Quantity? How will all the different units interact with this functionality?

mwcraig replied:
I think it is fine -- the PSF is just carried along sort of like the metadata. I don't see a reason to forbid `Quantity` given that. The uncertainties we need to pay attention because we do arithmetic with them but I think we are ok here with quantities without needing to force consistency of some sort between the units of the `CCDData` object and the unit (if any) of the `psf`.

pllim replied:
I am just worried that you will give wrong answer without realizing it if someone pass in Quantity with unit that somehow doesn't trip any exception. But if you are sure that is not a problem, then please change the test as you suggested. Thanks!

mwcraig replied:
Right now it isn't clear to me what unit to expect if a `Quantity` is passed in -- right now if any arithmetic is attempted with a `psf` attached we emit a warning.

Hopefully we get some user feedback on this after release.

pllim commented:
Thanks, all!

parejkoj commented:
Thank you for all the help in getting this merged!
